packages/controllers/src/permissions/Permission.ts

```diff
@@ -0,0 +1,462 @@
+  return permission.caveats?.find((caveat) => caveat.type === caveatType);
+}
+
+/**
+ * A requested permission object. Just an object with any of the properties
+ * of a {@link PermissionConstraint} object.
+ */
+type RequestedPermission = Partial<PermissionConstraint>;
+
+/**
+ * A record of target names and their {@link RequestedPermission} objects.
+ */
+export type RequestedPermissions = Record<
+  GenericTargetName,
+  RequestedPermission
+>;
+
+/**
+ * The restricted method context object. Essentially a way to pass internal
+ * arguments to restricted methods and caveat functions, most importantly the
+ * requesting origin.
+ */
+type RestrictedMethodContext = Readonly<{
+  origin: OriginString;
+  [key: string]: any;
+}>;
+
+export type RestrictedMethodParameters = Json[] | Record<string, Json> | void;
+
+/**
+ * The arguments passed to a restricted method implementation.
+ *
+ * @template Params - The JSON-RPC parameters of the restricted method.
+ */
+export type RestrictedMethodOptions<Params extends RestrictedMethodParameters> =
+  {
+    method: GenericTargetName;
+    params?: Params;
+    context: RestrictedMethodContext;
+  };
+
+/**
+ * A synchronous restricted method implementation.
+ *
+ * @template Params - The JSON-RPC parameters of the restricted method.
+ * @template Result - The JSON-RPC result of the restricted method.
+ */
+export type SyncRestrictedMethod<
+  Params extends RestrictedMethodParameters,
+  Result extends Json,
+> = (args: RestrictedMethodOptions<Params>) => Result;
+
+/**
+ * An asynchronous restricted method implementation.
+ *
+ * @template Params - The JSON-RPC parameters of the restricted method.
+ * @template Result - The JSON-RPC result of the restricted method.
+ */
+export type AsyncRestrictedMethod<
+  Params extends RestrictedMethodParameters,
+  Result extends Json,
+> = (args: RestrictedMethodOptions<Params>) => Promise<Result>;
+
+/**
+ * A synchronous or asynchronous restricted method implementation.
+ *
+ * @template Params - The JSON-RPC parameters of the restricted method.
+ * @template Result - The JSON-RPC result of the restricted method.
+ */
+export type RestrictedMethod<
+  Params extends RestrictedMethodParameters,
+  Result extends Json,
+> =
+  | SyncRestrictedMethod<Params, Result>
+  | AsyncRestrictedMethod<Params, Result>;
+
+export type GenericRestrictedMethod = RestrictedMethod<
+  RestrictedMethodParameters,
+  Json
+>;
+
+export type ValidRestrictedMethod<
+  MethodImplementation extends RestrictedMethod<any, any>,
+> = MethodImplementation extends (args: infer Options) => Json | Promise<Json>
+  ? Options extends RestrictedMethodOptions<RestrictedMethodParameters>
+    ? MethodImplementation
+    : never
+  : never;
+
+export type PermissionFactory<
+  TargetPermission extends PermissionConstraint,
+  RequestData extends Record<string, unknown>,
+> = (
+  options: PermissionOptions<TargetPermission>,
+  requestData?: RequestData,
+) => TargetPermission;
```

Comment:
Hmm. I was going to suggest dropping `TargetPermission` here as well, for similar reasons as with `constructPermission`. But this case is different. Here the factory function will be instantiated as part of a permission spec, with a specific permission in mind. Seems warranted.

The `PermissionOptions` generic isn't helping here though, so that can be dropped. It's still type safe without that.

Reply:
Oh sorry - I was referring to this suggestion https://github.com/MetaMask/snaps-skunkworks/pull/28#discussion_r744933877

The part about dropping the generic parameter from `PermissionOptions` applies here too. We can use the other thread to continue the discussion about that.